Adapt docs to PennyLane theme (#64)
PennyLaneAI/pennylane-forest: doc/devices/numpy_wavefunction.rst

The Numpy-Wavefunction device
=============================

The ``forest.numpy_wavefunction`` device provides an interface between PennyLane

> and the pyQuil `NumpyWavefunctionSimulator`.

Reviewer: Sorry, my bad from the previous iteration (needs double back ticks).
```suggestion
and the pyQuil ``NumpyWavefunctionSimulator``.
```

As the NumPy wavefunction simulator allows access and manipulation of the underlying
quantum state vector, ``forest.numpy_wavefunction`` is able to support the full
suite of PennyLane and Quil quantum operations and observables.

.. note::

>     Since the NumPy wavefunction simulator is written entirely in NumPy, no external
>     Quil compiler is required.
> 
> 

Reviewer 2: ```suggestion
    Since the NumPy wavefunction simulator is written entirely in NumPy, no external
    Quil compiler is required.

```

A trivial change to force a PR re-build.

The Numpy-Wavefunction device
=============================

> The ``forest.numpy_wavefunction`` device provides an interface between PennyLane
> and the pyQuil `NumPy wavefunction simulator <http://docs.rigetti.com/en/stable/wavefunction_simulator.html>`_.

Reviewer: The link seems to be broken here. This one seems to refer to the docs:
https://pyquil-docs.rigetti.com/en/v2.24.0/apidocs/autogen/pyquil.numpy_simulator.NumpyWavefunctionSimulator.html

However, it includes the version number and could otherwise be subject to a change of URL eventually breaking the link again. Maybe good to go without a link here?
```suggestion
The ``forest.numpy_wavefunction`` device provides an interface between PennyLane
and the pyQuil `NumpyWavefunctionSimulator`.
```

> Because the NumPy wavefunction simulator allows access and manipulation of the underlying

Reviewer: ```suggestion
As the NumPy wavefunction simulator allows access and manipulation of the underlying
```

quantum state vector, ``forest.numpy_wavefunction`` is able to support the full

> suite of PennyLane and Quil quantum operations and observables.

Reviewer: :100:

In addition, it is generally faster than running equivalent simulations on the QVM, as the

> final state can be inspected and the expectation value calculated analytically, rather than by sampling measurements.

Reviewer: Hmm, not quite sure I agree here. From my experience, the main reason for slower execution with the QVM is that a local Lisp server is used that accepts/sends queries. Furthermore, the `forest.numpy_wavefunction` device could also be run in `analytic=False` mode.

Author: Deleted that statement.

.. note::

    Since the NumPy wavefunction simulator is written entirely in NumPy, no external
    Quil compiler is required.

.. note::

    By default, ``forest.numpy_wavefunction`` is initialized with ``shots=0``, indicating

>     that the exact analytic expectation value is to be returned.

Reviewer: Wait, was the `analytic=True` == `shots=0` change made in PennyLane core? :thinking: Might have missed it

Author: I changed it to `analytic=True`, but this goes to show that we should check the device docs when releasing...

Reviewer 2: @antalszava I think this block of text is so old, it is referring to the pre PL v0.3 way of doing things 😆

    instead non-zero, a spectral decomposition is performed and a Bernoulli distribution
    is constructed and sampled. This allows the ``forest.numpy_wavefunction`` device to
    'approximate' the effect of sampling the expectation value.

Usage
~~~~~

You can instantiate the device in PennyLane as follows:

.. code-block:: python

    import pennylane as qml

    dev_numpy = qml.device('forest.numpy_wavefunction', wires=2)

This device can then be used just like other devices for the definition and evaluation of QNodes within PennyLane.

A simple quantum function that returns the expectation value and variance of a measurement and
depends on three classical input parameters would look like:

.. code-block:: python

    @qml.qnode(dev_numpy)
    def circuit(x, y, z):
        qml.RZ(z, wires=[0])
        qml.RY(y, wires=[0])
        qml.RX(x, wires=[0])
        qml.CNOT(wires=[0, 1])
        return qml.expval(qml.PauliZ(0)), var(qml.PauliZ(1))

You can then execute the circuit like any other function to get the quantum mechanical expectation value and variance:

>>> circuit(0.2, 0.1, 0.3)
array([0.97517033, 0.04904283])

Supported operations
~~~~~~~~~~~~~~~~~~~~

> All Forest devices support all PennyLane `operations and observables <https://pennylane.readthedocs.io/en/latest/code/ops/qubit.html>`_, with

Reviewer 3: ```suggestion
All Forest devices support all PennyLane `operations and observables <https://pennylane.readthedocs.io/en/stable/introduction/operations.html#qubit-operations>`_, with
```
